File: Backend/Microservices/Expense/EMS.Core/Entities/RecurrenceRule.cs
using EMS.Core.Entities.Common;
using EMS.Core.Enums;

namespace EMS.Core.Entities
{
    public class RecurrenceRule : BaseAuditableEntity<int>
    {
        public RecurrenceType Frequency { get; set; }
        private int _interval = 1;
        public int Interval { get => _interval; set => _interval = value > 0 ? value : 1; }
        public string[] ByDayOfWeek { get; set; } = [];
        public int[] ByMonthDay { get; set; } = [];
        public int[] ByMonth { get; set; } = [];
        public DateTimeOffset? EndDate { get; set; }
        private int? _maxOccurrences = default;
        public int? MaxOccurrences { get => _maxOccurrences; set => _maxOccurrences = value == null 
                ? null 
                : value > 0 
                    ? value 
                    : _maxOccurrences; }

        // Navigations
    }
}

File: Backend/Microservices/AI/app/schemas/recurrence_rule.py
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field

from app.enums.recurrence_type import RecurrenceType


class RecurrenceRule(BaseModel):
    """Recurrence rule for scheduling tasks.
    Follows iCalendar specification.
    """

    frequency: RecurrenceType = Field(
        description="Frequency of recurrence. MUST be one of: 'Daily', 'Weekly', 'Monthly', 'Yearly'.",
    )
    interval: Optional[int] = Field(
        default=1,
        description="Interval of the frequency (e.g., 2 for every 2 weeks). Defaults to 1.",
    )
    by_day_of_week: Optional[List[str]] = Field(
        default=None,
        description="List of ISO days ('MO', 'TU', 'WE', 'TH', 'FR', 'SA', 'SU'). Used for Weekly, Monthly, Yearly.",
    )
    by_month_day: Optional[List[int]] = Field(
        default=None,
        description="List of days of the month (1-31) or negative indices (-1 for last, -2 for second last). Used for Monthly, Yearly.",
    )
    by_month: Optional[List[int]] = Field(
        default=None, description="List of months (1-12). Used for Yearly."
    )
    end_date: Optional[datetime] = Field(
        default=None, description="The UTC datetime after which the recurrence stops."
    )
    max_occurrences: Optional[int] = Field(
        default=None, description="Maximum number of times the event should occur."
    )
